modules/ocha_ai_chat/src/Form/OchaAiChatChatForm.php

```diff
@@ -169,7 +169,13 @@ public function buildForm(array $form, FormStateInterface $form_state, ?bool $po
       ];
     }
 
+    // Read config to determine feedback type for each history entry.
+    $feedback_type = \Drupal::config('ocha_ai_chat.settings')->get('defaults.form.feedback');
```

Comment:
gotcha. Just out of curiosity, is that a piece of institutional knowledge of `FormBase` that I just have to know, or is there a way to look at the code and know somehow?

Reply:
@rupl Usually, when I need to use a service in a class, I look at the base classes to see if they are already injected or available through this type of methods.

`::config()`, `::logger()`, `::getEntityTypeManager()` are common ones present in ControllerBase, FormBase etc.

If not present, you usually need to add 4/5 things:
1.  Lookup the service in https://api.drupal.org/api/drupal/core%21core.services.yml/8.2.x
2. Take note of the class and open the Drupal API page for it.
3. See if it has an interface, note its namespaced path otherwise note the class path 
4. Add `use XXX\YYY` (interface path or class path) at the top of your file (ex: OchaAIChatForm.php)
5. Add a new variable to the class to store the service (`protected ServiceInterface $service`)
6. Add a parameter to the constructor with (`__construct(..., ServiceInterface $service) {... $this->service = $service; }`), update the constructor comment with the new parameter.
7. Add the service in the static `create(...) { return new static(..., $container->get('service')); }` method if there is one (for services, add the dependency in `mymodule.services.yml)

For some services, there are some traits you can directly that will provide a method to get the service.

Finally, with recent Drupal and PHP versions you can combine 5 and 6 in something like  `__construct(..., protected ServiceInterface $service) { ... }` which is a shortcut (ex: https://github.com/UN-OCHA/rwint9-site/blob/e494d86419b297c6bdf60aeba5814c9fdf2ae317/html/modules/custom/reliefweb_post_api/src/Plugin/ContentProcessorPluginBase.php#L106. This seems to be the new recommended moving forward as it reduces a bit the code (no variable declaration in class and no `$this->bidule = $bidule` in constructor).

Reply:
Thanks for the explainer!